packages/unified-selection/src/unified-selection/HiliteSetCache.ts

```diff
@@ -0,0 +1,98 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+/** @packageDocumentation
+ * @module UnifiedSelection
+ */
+
+import { from, Observable, shareReplay } from "rxjs";
+import { eachValueFrom } from "rxjs-for-await";
+import { HiliteSet, HiliteSetProvider } from "./HiliteSetProvider";
+import { IMetadataProvider } from "./queries/ECMetadata";
+import { IECSqlQueryExecutor } from "./queries/ECSqlCore";
+import { StorageSelectionChangeEventArgs } from "./SelectionChangeEvent";
+import { SelectionStorage } from "./SelectionStorage";
+
+/**
+ * Caches hilite set for the current `SelectionStorage` selection
+ * so subsequent requests for the same input does not cost
+ * @beta
+ */
+export interface HiliteSetCache {
+  /** Get the current hilite set iterator for the specified imodel */
+  getHiliteSet(props: {
+    /** iModel to get hilite set for */
+    iModelKey: string;
+    /** ECSql query executor */
+    queryExecutor: IECSqlQueryExecutor;
+    /** EC metadata provider */
+    metadataProvider: IMetadataProvider;
+  }): AsyncIterableIterator<HiliteSet>;
+
+  /** Clears hilite set cache for a given iModel */
+  clearCache(props: {
+    /** Key of the iModel to change selection for */
+    iModelKey: string;
+  }): void;
+
+  /** Disposes the cache. */
+  dispose(): void;
+}
```

Comment:
I'd like the API of this to be similar to `HiliteSetProvider`:
```ts
interface CachingHiliteSetProvider extends HiliteSetProvider {
    dispose(): void;
}
```

I understand we need query executor and metadata provider, which we won't get on selection changes, but I think we could request a getter in the "create" function, e.g.:
```ts
function createCachingSelectionHiliteSetProvider(props: {
    selectionStorage: SelectionStorage;
    iModelProvider: (iModelKey: string): { queryExecutor: IECSqlQueryExecutor; metadataProvider: IMetadataProvider };
}): CachingHiliteSetProvider {
    // return slightly modified HiliteSetCacheImpl
}
```

Reply:
I missed one part of my suggestion:
```ts
interface HiliteSetProvider {
  getHiliteSet(props: { iModelKey: string }): AsyncIterableIterator<HiliteSet>;
}
```

- Existing `HiliteSetProvider` class becomes `HiliteSetProviderImpl`.
- `HiliteSetProvider.create` becomes a root level `createHiliteSetProvider`.

Reply:
The existing `getHiliteSet` in `HiliteSetProvider` requires `Selectables` as argument rather than `iModelKey`, while the `CachingHiliteSetProvider` requires `iModelKey`.

Reply:
Oh, okay - I missed that piece... yeah, sending selectables to this - based on selection - one doesn't make sense. 

What I'd like to avoid is consumers having to pass those 3 arguments to get a hilite set - that's going to be very inconvenient... `iModelKey` is okay, but the other two aren't.

Reply:
Refactored to not need to pass the `queryExecutor` and `metadataProvider` every call.